@tdonohue there are tests that explicitly reference taskFiles from the processes interface and expect them to work: 
```
Error:  Tests run: 27, Failures: 6, Errors: 15, Skipped: 0, Time elapsed: 54.02 s <<< FAILURE! -- in org.dspace.curate.CurationScriptIT
Error:  org.dspace.curate.CurationScriptIT.curateScript_validRequest_TaskFile -- Time elapsed: 0.619 s <<< FAILURE!
java.lang.AssertionError: 
JSON path "$"
```

So that means we've also set user expectations that this is possible too. My feeling is we should leave it in for v10 and rely on the stricter "allowed paths" config, rather than break anything that relies on admins being able to run queues/task files from the UI. Thoughts? If we remove it we'll have to update the tests and make sure everyone is aware that this functionality no longer works.

For now, I have force pushed this branch with that commit dropped, so we get passing tests. I have a backup commit stored if we decide to add it back in and address tests/documentation separately.

--- a/dspace-api/src/main/java/org/dspace/core/LDN.java
+++ b/dspace-api/src/main/java/org/dspace/core/LDN.java
@@ -10,14 +10,18 @@ package org.dspace.core;
 import static org.apache.commons.lang3.StringUtils.EMPTY;
 
 import java.io.BufferedReader;
+import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.io.StringWriter;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
+import java.util.Properties;
 
 import jakarta.mail.MessagingException;
 import org.apache.commons.lang3.StringUtils;
@@ -25,11 +29,15 @@ import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.velocity.Template;
 import org.apache.velocity.VelocityContext;
+import org.apache.velocity.app.Velocity;
 import org.apache.velocity.app.VelocityEngine;
 import org.apache.velocity.exception.MethodInvocationException;
 import org.apache.velocity.exception.ParseErrorException;
 import org.apache.velocity.exception.ResourceNotFoundException;
+import org.apache.velocity.runtime.resource.loader.StringResourceLoader;
 import org.apache.velocity.runtime.resource.util.StringResourceRepository;
+import org.dspace.services.ConfigurationService;
+import org.dspace.services.factory.DSpaceServicesFactory;
 
 /**
  * Class representing an LDN message json
@@ -52,9 +60,29 @@ public class LDN {
 
     /** Velocity template settings. */
     private static final String RESOURCE_REPOSITORY_NAME = "LDN";
+    private static final Properties VELOCITY_PROPERTIES = new Properties();
+    static {
+        VELOCITY_PROPERTIES.put(Velocity.RESOURCE_LOADERS, "string");
+        VELOCITY_PROPERTIES.put("resource.loader.string.description",
+                "Velocity StringResource loader");
+        VELOCITY_PROPERTIES.put("resource.loader.string.class",
+                StringResourceLoader.class.getName());
+        VELOCITY_PROPERTIES.put("resource.loader.string.repository.name",
+                RESOURCE_REPOSITORY_NAME);
+        VELOCITY_PROPERTIES.put("resource.loader.string.repository.static",
+                "false");
+    }
 
     /** Velocity template for the message*/
     private Template template;
+
+    /** Allowed base directory for LDN messages / templates **/
+    private static final ConfigurationService configurationService =
+        DSpaceServicesFactory.getInstance().getConfigurationService();
+    private static final String dspaceDir = configurationService.getProperty("dspace.dir", "/dspace");
+    private static final Path allowedTemplateBasePath = Paths.get(
+            configurationService.getProperty( "ldn.template.path", dspaceDir + File.separatorChar
+                    + "config" + File.separatorChar + "ldn"));
 
     /**
      * Create a new ldn message.
@@ -95,6 +123,9 @@ public class LDN {
      * @throws IOException        if IO error
      */
     public String generateLDNMessage() {
+        ConfigurationService config
+            = DSpaceServicesFactory.getInstance().getConfigurationService();
+
         VelocityEngine templateEngine = new VelocityEngine();
         templateEngine.init(Utils.getSecureVelocityProperties(RESOURCE_REPOSITORY_NAME));
 
@@ -144,8 +175,14 @@ public class LDN {
     public static LDN getLDNMessage(String ldnMessageFile)
         throws IOException {
         StringBuilder contentBuffer = new StringBuilder();
+        Path ldnMessagePath = new File(ldnMessageFile).toPath().normalize();
+        Path realLdnMessagePath = allowedTemplateBasePath.resolve(ldnMessagePath).normalize();
+        if (!realLdnMessagePath.startsWith(allowedTemplateBasePath)) {
+            throw new IOException("Illegal LDN message path: '" + ldnMessagePath + "'");
+        }
+
         try (
-            InputStream is = new FileInputStream(ldnMessageFile);
+            InputStream is = new FileInputStream(realLdnMessagePath.toFile());
             InputStreamReader ir = new InputStreamReader(is, "UTF-8");
             BufferedReader reader = new BufferedReader(ir);
             ) {

--- a/dspace-api/src/main/java/org/dspace/curate/Curation.java
+++ b/dspace-api/src/main/java/org/dspace/curate/Curation.java
@@ -10,12 +10,12 @@ package org.dspace.curate;
 import java.io.BufferedReader;
 import java.io.File;
 import java.io.FileNotFoundException;
-import java.io.FileReader;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.io.OutputStreamWriter;
 import java.io.PrintStream;
 import java.io.Writer;
+import java.nio.charset.StandardCharsets;
 import java.sql.SQLException;
 import java.time.Instant;
 import java.util.HashMap;
@@ -39,6 +39,7 @@ import org.dspace.handle.factory.HandleServiceFactory;
 import org.dspace.handle.service.HandleService;
 import org.dspace.scripts.DSpaceRunnable;
 import org.dspace.services.factory.DSpaceServicesFactory;
+import org.dspace.storage.secure.SecureFileAccess;
 import org.dspace.utils.DSpace;
 
 /**
@@ -114,7 +115,12 @@ public class Curation extends DSpaceRunnable<CurationScriptConfiguration> {
             // load taskFile
             BufferedReader reader = null;
             try {
-                reader = new BufferedReader(new FileReader(this.taskFile));
+                String dspaceDir = DSpaceServicesFactory.getInstance()
+                    .getConfigurationService().getProperty("dspace.dir");
+                String allowedTaskFileBasePath = DSpaceServicesFactory.getInstance()
+                    .getConfigurationService().getProperty("curate.taskfile.base", dspaceDir);
+                reader = SecureFileAccess.getBufferedReader(this.taskFile, allowedTaskFileBasePath,
+                        "curation-taskfile", StandardCharsets.UTF_8);
                 while ((taskName = reader.readLine()) != null) {
                     if (verbose) {
                         super.handler.logInfo("Adding task: " + taskName);
@@ -190,12 +196,23 @@ public class Curation extends DSpaceRunnable<CurationScriptConfiguration> {
     private Curator initCurator() throws FileNotFoundException {
         Curator curator = new Curator(handler);
         OutputStream reporterStream;
+        String dspaceDir = DSpaceServicesFactory.getInstance()
+            .getConfigurationService().getProperty("dspace.dir");
+        String allowedReporterBasePath = DSpaceServicesFactory.getInstance()
+            .getConfigurationService().getProperty("curate.reporter.base",
+                    dspaceDir + File.separatorChar + "log");
         if (null == this.reporter) {
-            reporterStream = NullOutputStream.NULL_OUTPUT_STREAM;
+            reporterStream = NullOutputStream.INSTANCE;
         } else if ("-".equals(this.reporter)) {
             reporterStream = System.out;
         } else {
-            reporterStream = new PrintStream(this.reporter);
+            try {
+                reporterStream = new PrintStream(
+                    SecureFileAccess.getOutputStream(
+                    this.reporter, allowedReporterBasePath, "curation-reporter"));
+            } catch (IOException e) {
+                throw new FileNotFoundException(e.getLocalizedMessage());
+            }
         }
         Writer reportWriter = new OutputStreamWriter(reporterStream);
         curator.setReporter(reportWriter);

--- a/dspace-api/src/main/java/org/dspace/storage/secure/SecureFileAccess.java
+++ b/dspace-api/src/main/java/org/dspace/storage/secure/SecureFileAccess.java
@@ -14,7 +14,6 @@ import java.io.OutputStream;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
-import java.nio.file.LinkOption;
 import java.nio.file.Path;
 import java.util.List;
 
@@ -62,7 +61,7 @@ public final class SecureFileAccess {
      * More secure than the 'write' variant because we can explicitly resolve links as well.
      *
      * @param file the unvalidated file, usually derived from user input or configuration
-     * @param allowedBasePath the allowed base path for this use case as per system configuration
+     * @param allowedBasePaths the allowed base paths for this use case as per system configuration
      * @param purpose the name of the calling component / use case for logging and inspection
      * @throws IOException on validation failure
      */
@@ -72,9 +71,6 @@ public final class SecureFileAccess {
             Path basePath = Path.of(allowedBasePath)
                                 .toRealPath()
                                 .normalize();
-            if (basePath == null) {
-                throw new IOException("Null base path can not be provided for validation");
-            }
             Path resolvedPath = basePath.resolve(file).toRealPath().normalize();
             if (resolvedPath.startsWith(basePath)) {
                 return resolvedPath;
@@ -116,8 +112,8 @@ public final class SecureFileAccess {
     }
 
     /**
-     * Get an input stream after validating file path. Adds NOFOLLOW_LINKS link option to
-     * help ensure some extra safety since new files can't use toRealPath() for link calculation
+     * Get an output stream after validating file path. New files can't use toRealPath() for link calculation so
+     * there is a bit of a trade-off in allowing some symlink traversal to occur
      * @param unvalidatedFile the unvalidated file, usually derived from user input or configuration
      * @param allowedBasePaths the allowed base paths for this use case as per system configuration
      * @param purpose the name of the calling component / use case for logging and inspection
@@ -126,6 +122,6 @@ public final class SecureFileAccess {
     public static OutputStream getOutputStream(String unvalidatedFile, List<String> allowedBasePaths, String purpose)
             throws IOException {
         Path validatedFile = validatePathForWrite(unvalidatedFile, allowedBasePaths, purpose);
-        return Files.newOutputStream(validatedFile, LinkOption.NOFOLLOW_LINKS);
+        return Files.newOutputStream(validatedFile);
     }
 }

--- a/dspace-api/src/main/java/org/dspace/storage/secure/SecureFileAccess.java
+++ b/dspace-api/src/main/java/org/dspace/storage/secure/SecureFileAccess.java
@@ -1,0 +1,121 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.storage.secure;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+
+/**
+ * Decent I/O path validation - not perfect when symlinks are used and we are writing
+ * as 'toRealPath' check on the resolved path fails for new files
+ *
+ * @author Kim Shepherd <kim-at-shepherd-dot-nz>
+ */
+public final class SecureFileAccess {
+
+    private SecureFileAccess() {}
+
+    /**
+     * Validate a given path against an allowed base path. Does not attempt to calculate "real path"
+     * before validation, as this breaks for new files which don't yet exist. This can make the resulting
+     * validation still vulnerable to symlink traversal in some cases
+     * @param file the unvalidated file, usually derived from user input or configuration
+     * @param allowedBasePath the allowed base path for this use case as per system configuration
+     * @param purpose the name of the calling component / use case for logging and inspection
+     * @throws IOException on validation failure
+     */
+    public static Path validatePathForWrite(String file, String allowedBasePath, String purpose)
+            throws IOException {
+        Path basePath = Path.of(allowedBasePath)
+                            .toRealPath()
+                            .normalize();
+        if (basePath == null) {
+            throw new IOException("Null base path can not be provided for validation");
+        }
+        Path resolvedPath = basePath.resolve(file).normalize();
+        if (!resolvedPath.startsWith(basePath)) {
+            throw new IOException("Illegal file path attempted for I/O (%s): %s".formatted(purpose, file));
+        }
+        return resolvedPath;
+    }
+
+    /**
+     * Validate a given path against an allowed base path.
+     * More secure than the 'write' variant because we can explicitly resolve links as well.
+     *
+     * @param file the unvalidated file, usually derived from user input or configuration
+     * @param allowedBasePath the allowed base path for this use case as per system configuration
+     * @param purpose the name of the calling component / use case for logging and inspection
+     * @throws IOException on validation failure
+     */
+    public static Path validatePathForRead(String file, String allowedBasePath, String purpose)
+            throws IOException {
+        Path basePath = Path.of(allowedBasePath)
+                            .toRealPath()
+                            .normalize();
+        if (basePath == null) {
+            throw new IOException("Null base path can not be provided for validation");
+        }
+        Path resolvedPath = basePath.resolve(file).toRealPath().normalize();
+        if (!resolvedPath.startsWith(basePath)) {
+            throw new IOException("Illegal file path attempted for I/O (%s): %s".formatted(purpose, file));
+        }
+        return resolvedPath;
+    }
+
+    /**
+     * Get a buffered reader after validating file path.
+     * @param unvalidatedFile the unvalidated file, usually derived from user input or configuration
+     * @param allowedBasePath the allowed base path for this use case as per system configuration
+     * @param purpose the name of the calling component / use case for logging and inspection
+     * @throws IOException on validation failure
+     */
+    public static BufferedReader getBufferedReader(String unvalidatedFile, String allowedBasePath,
+            String purpose, Charset charset) throws IOException {
+        if (charset == null) {
+            charset = StandardCharsets.UTF_8;
+        }
+        Path validatedFile = validatePathForRead(unvalidatedFile, allowedBasePath, purpose);
+        return Files.newBufferedReader(validatedFile, charset);
+    }
+
+    /**
+     * Get an input stream after validating file path.
+     * @param unvalidatedFile the unvalidated file, usually derived from user input or configuration
+     * @param allowedBasePath the allowed base path for this use case as per system configuration
+     * @param purpose the name of the calling component / use case for logging and inspection
+     * @throws IOException on validation failure
+     */
+    public static InputStream getInputStream(String unvalidatedFile, String allowedBasePath, String purpose)
+            throws IOException {
+        Path validatedFile = validatePathForRead(unvalidatedFile, allowedBasePath, purpose);
+        return Files.newInputStream(validatedFile);
+
+    }
+
+    /**
+     * Get an input stream after validating file path. Adds NOFOLLOW_LINKS link option to
+     * help ensure some extra safety since new files can't use toRealPath() for link calculation
+     * @param unvalidatedFile the unvalidated file, usually derived from user input or configuration
+     * @param allowedBasePath the allowed base path for this use case as per system configuration
+     * @param purpose the name of the calling component / use case for logging and inspection
+     * @throws IOException on validation failure
+     */
+    public static OutputStream getOutputStream(String unvalidatedFile, String allowedBasePath, String purpose)
+            throws IOException {
+        Path validatedFile = validatePathForWrite(unvalidatedFile, allowedBasePath, purpose);
+        return Files.newOutputStream(validatedFile, LinkOption.NOFOLLOW_LINKS);
+    }
+}

--- a/dspace-api/src/test/java/org/dspace/app/ldn/action/SendLDNMessageActionIT.java
+++ b/dspace-api/src/test/java/org/dspace/app/ldn/action/SendLDNMessageActionIT.java
@@ -10,10 +10,14 @@ package org.dspace.app.ldn.action;
 import static org.dspace.app.ldn.action.LDNActionStatus.ABORT;
 import static org.dspace.app.ldn.action.LDNActionStatus.CONTINUE;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
 import static org.mockito.Mockito.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
+import java.io.IOException;
 import java.io.InputStream;
 import java.sql.SQLException;
 import java.util.List;
@@ -40,6 +44,7 @@ import org.dspace.builder.WorkspaceItemBuilder;
 import org.dspace.content.Collection;
 import org.dspace.content.Item;
 import org.dspace.content.WorkspaceItem;
+import org.dspace.core.LDN;
 import org.dspace.eperson.EPerson;
 import org.dspace.services.ConfigurationService;
 import org.dspace.services.factory.DSpaceServicesFactory;
@@ -230,6 +235,28 @@ public class SendLDNMessageActionIT extends AbstractIntegrationTestWithDatabase 
         assertEquals(sendLDNMessageAction.execute(context, notification, item), ABORT);
         mockedClient.close();
         response.close();
+    }
+
+    @Test
+    public void testLDNLegalPath() throws Exception {
+        try {
+            // Path traversal will be calculated properly
+            // but this still ends up at a legal base
+            LDN message = LDN.getLDNMessage("../../config/ldn/request-review");
+            assertNotNull(message);
+        } catch (IOException e) {
+            fail("IOException should NOT have been thrown for legal template path: " + e.getMessage());
+        }
+    }
+
+    @Test
+    public void testLDNIllegalPath() throws Exception {
+        try {
+            LDN.getLDNMessage("../modules/dspace.cfg");
+            fail("IOException should have been thrown for illegal template path");
+        } catch (IOException e) {
+            assertTrue(e.getMessage().contains("Illegal LDN message path:"));
+        }
     }
 
     @Override

--- a/dspace/config/modules/curate.cfg
+++ b/dspace/config/modules/curate.cfg
@@ -30,3 +30,12 @@ curate.taskqueue.dir = ${dspace.dir}/ctqueues
 
 # Maximum amount of redirects set to 0 for none and -1 for unlimited
 curate.checklinks.max-redirect = 0
+
+# allowed base path of curation task files
+# it is recommended to restrict this path as much as possible
+# so that the DSpace Processes framework may only load files as "tasks"
+# from a trusted location
+curate.taskfile.base = ${dspace.dir}
+
+# allowed base path of reporter output.
+curate.reporter.base = ${dspace.dir}/log

--- a/dspace/config/modules/ldn.cfg
+++ b/dspace/config/modules/ldn.cfg
@@ -40,6 +40,11 @@ ldn.notify.inbox.block-untrusted-ip = true
 # this is the medatada used to retrieve the relation with external items when sending relationship requests
 #ldn.notify.relation.metadata = dc.relation
 
+# Allowed base path of LDN templates. Apache Velocity will only
+# load templates that begin with this base path.
+# Default is ${dspace.dir}/config/ldn
+#ldn.template.path = ${dspace.dir}/config/ldn
+
 
 # EMAIL CONFIGURATION
 # Supported values for actionSendFilter are:

